src/test/java/io/cdap/plugin/aws/s3/source/S3BatchSourceTest.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package io.cdap.plugin.aws.s3.source;
+
+import io.cdap.cdap.etl.mock.validation.MockFailureCollector;
+import io.cdap.plugin.aws.s3.common.S3ConnectorConfig;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+
+/**
+ * Test class for S3 Batch Source
+ */
+public class S3BatchSourceTest {
+
+  @Test
+  public void testValidateWithoutCredentials() throws IllegalAccessException, NoSuchFieldException {
+    MockFailureCollector mockFailureCollector = new MockFailureCollector();
+    S3BatchSource.S3BatchConfig s3BatchConfig = new S3BatchSource.S3BatchConfig();
+    Field referenceNameField = s3BatchConfig.getClass().getSuperclass().getDeclaredField("referenceName");
+    referenceNameField.setAccessible(true);
+    referenceNameField.set(s3BatchConfig, "ref");
+    Field pathField = s3BatchConfig.getClass().getDeclaredField("path");
+    pathField.setAccessible(true);
+    pathField.set(s3BatchConfig, "s3a://my-bucket/part1");
+    s3BatchConfig.validate(mockFailureCollector);
+    Assert.assertEquals(1, mockFailureCollector.getValidationFailures().size());
+    Assert.assertEquals("Connection credentials is not provided",
+                        mockFailureCollector.getValidationFailures().get(0).getMessage());
+  }
+
+  @Test
+  public void testValidateWithInvalidCredentials() throws IllegalAccessException, NoSuchFieldException {
+    MockFailureCollector mockFailureCollector = new MockFailureCollector();
+    S3BatchSource.S3BatchConfig s3BatchConfig = new S3BatchSource.S3BatchConfig();
+    S3ConnectorConfig s3ConnectorConfig = new S3ConnectorConfig("accessID", "accessKey", "sessionToken",
+      "Access Credentials", "region");
+    Field connectionField = s3BatchConfig.getClass().getDeclaredField("connection");
+    connectionField.setAccessible(true);
+    connectionField.set(s3BatchConfig, s3ConnectorConfig);
+    Field referenceNameField = s3BatchConfig.getClass().getSuperclass().getDeclaredField("referenceName");
+    referenceNameField.setAccessible(true);
+    referenceNameField.set(s3BatchConfig, "ref");
+    Field pathField = s3BatchConfig.getClass().getDeclaredField("path");
+    pathField.setAccessible(true);
+    pathField.set(s3BatchConfig, "s3a://my-bucket/part1");
+    s3BatchConfig.validate(mockFailureCollector);
```

Comment:
This one is ideally not unit test as it is taking dependency on external system. Ideal would have been to just test the "getS3Client" and "validatePath" function in isolation. Since validateCredentials, makes an external call, it is not possible to keep it in unit test category.

Reply:
Yes, it's more of an integration test. This repo contains two other test classes ([S3PathTest.java](https://github.com/data-integrations/amazon-s3-plugins/blob/develop/src/test/java/io/cdap/plugin/aws/s3/common/S3PathTest.java) and   [S3ConnectorTest.java](https://github.com/data-integrations/amazon-s3-plugins/blob/develop/src/test/java/io/cdap/plugin/aws/s3/connector/S3ConnectorTest.java)) having a mix of both unit as well as integration tests.

In order to test the `getS3Client()` and `validatePath()` separately, I'll need to change their access modifier from _private_ to _package-private_ and make them visible for testing. While this can be done, changing their scope is in conflict with the earlier [suggestion](https://github.com/data-integrations/amazon-s3-plugins/pull/185/#discussion_r1340938860) to make the methods private.

Reply:
It is okay to make the methods `package-private` and add the annotation `@VisibleForTesting` we have done it in few other places.

Make sure we add an e2e test in a follow up PR.